File: data_structure/binaryTree/inorder/inoder_traversal_items.py

# 二叉树的中序遍历

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# 递归
def inorderTraversal(root: TreeNode) ->[int]:
    result = []
    if not root:
        return []
    left = inorderTraversal(root.left)
    if left:
        result.extend(left)
    result.append(root.val)
    right = inorderTraversal(root.right)
    result.extend(right)
    return result


if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(2)

    root.left.left = TreeNode(3)
    root.left.right = TreeNode(4)

    root.right.left = TreeNode(4)
    root.right.right = TreeNode(3)
    result = inorderTraversal(root)
    print(result)

File: data_structure/binaryTree/inorder/inoder_traversal_items.go
package main

import (
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func main() {

	root := &TreeNode{Val: 1, Left: &TreeNode{Val: 2}, Right: &TreeNode{Val: 2}}
	root.Left.Left = &TreeNode{Val: 3}
	root.Left.Right = &TreeNode{Val: 4}
	root.Right.Left = &TreeNode{Val: 4}
	root.Right.Right = &TreeNode{Val: 3}
	result := inorderTraversal(root)
	fmt.Println(result)

}

// 迭代 中序遍历二叉树
func inorderTraversal(root *TreeNode) []int {
	result := []int{}
	queue := []*TreeNode{}
	for root != nil || len(queue) != 0 {
		for root != nil {
			queue = append(queue, root)
			root = root.Left
		}
		root = queue[len(queue)-1]
		queue = queue[:len(queue)-1]
		result = append(result, root.Val)
		root = root.Right
	}
	return result
}
